packages/nouns-contracts/contracts/NounsDescriptorV3.sol

```diff
@@ -0,0 +1,618 @@
+            return dataURI(tokenId, seed);
+        }
+        return string(abi.encodePacked(baseURI, tokenId.toString()));
+    }
+
+    /**
+     * @notice Given a token ID and seed, construct a base64 encoded data URI for an official Nouns DAO noun.
+     */
+    function dataURI(uint256 tokenId, INounsSeeder.Seed memory seed) public view override returns (string memory) {
+        string memory nounId = tokenId.toString();
+        string memory name = string(abi.encodePacked('Noun ', nounId));
+        string memory description = string(abi.encodePacked('Noun ', nounId, ' is a member of the Nouns DAO'));
+
+        return genericDataURI(name, description, seed);
+    }
+
+    /**
+     * @notice Given a name, description, and seed, construct a base64 encoded data URI.
+     */
+    function genericDataURI(
+        string memory name,
+        string memory description,
+        INounsSeeder.Seed memory seed
+    ) public view override returns (string memory) {
+        NFTDescriptorV2.TokenURIParams memory params = NFTDescriptorV2.TokenURIParams({
+            name: name,
+            description: description,
+            parts: getPartsForSeed(seed),
+            background: art.backgrounds(seed.background)
+        });
+        return NFTDescriptorV2.constructTokenURI(renderer, params);
+    }
+
+    /**
+     * @notice Given a seed, construct a base64 encoded SVG image.
+     */
+    function generateSVGImage(INounsSeeder.Seed memory seed) external view override returns (string memory) {
+        ISVGRenderer.SVGParams memory params = ISVGRenderer.SVGParams({
+            parts: getPartsForSeed(seed),
+            background: art.backgrounds(seed.background)
+        });
+        return NFTDescriptorV2.generateSVGImage(renderer, params);
+    }
+
+    /**
+     * @notice Get all Noun parts for the passed `seed`.
+     */
+    function getPartsForSeed(INounsSeeder.Seed memory seed) public view returns (ISVGRenderer.Part[] memory) {
+        bytes memory body = art.bodies(seed.body);
+        bytes memory accessory = art.accessories(seed.accessory);
+        bytes memory head = art.heads(seed.head);
+        bytes memory glasses_ = art.glasses(seed.glasses);
+
+        ISVGRenderer.Part[] memory parts = new ISVGRenderer.Part[](4);
+        parts[0] = ISVGRenderer.Part({ image: body, palette: _getPalette(body) });
+        parts[1] = ISVGRenderer.Part({ image: accessory, palette: _getPalette(accessory) });
+        parts[2] = ISVGRenderer.Part({ image: head, palette: _getPalette(head) });
+        parts[3] = ISVGRenderer.Part({ image: glasses_, palette: _getPalette(glasses_) });
+        return parts;
+    }
+
+    /**
+     * @notice Get the color palette pointer for the passed part.
+     */
+    function _getPalette(bytes memory part) private view returns (bytes memory) {
+        return art.palettes(uint8(part[0]));
+    }
+
+    /**
+     * @notice Replace current batch of accessories images with new ones.
```

Comment:
@EricDevito similar to my comment in the art file, seems important to clarify that we're replacing the entire trait, not just a "current batch".